### PR TITLE
New commands, bug fixes and support for aliases

### LINKED ISSUE #2: Add status command
Should pretty-print the statuses of all repos.

We need to figure out which information are the most important and a nice way to print them (some sort of a table?).

### LINKED ISSUE #45: Known errors should not be logged as crashes
image

I may be the author of this, but I was confused whether the tool failed gracefully or heavily. If we catch an issue, it shouldn't be that verbose. We just need the message to be logged.

### LINKED ISSUE #49: Unhandled promise: TypeError: msg.trim is not a function
```
 #  @ckeditor/ckeditor5-theme-lark                                       16/20 (80%)
Already on 'master'
Your branch is up-to-date with 'origin/master'.
From github.com:ckeditor/ckeditor5-theme-lark
 * branch            master     -> FETCH_HEAD
Current branch master is up to date.

{ logs:
   { error: [ 'From github.com:ckeditor/ckeditor5-utils\n * branch            master     -> FETCH_HEAD\n\nIt seems that there is already a rebase-apply directory, and\nI wonder if you are in the middle of another rebase.  If that is the\ncase, please try\n\tgit rebase (--continue | --abort | --skip)\nIf that is not the case, please\n\trm -fr "/workspace/ckeditor5/packages/ckeditor5-utils/.git/rebase-apply"\nand run me again.  I am stopping in case you still have something\nvaluable there.' ],
     info: [] } }
TypeError: msg.trim is not a function
    at Object.log (/workspace/mgit2/lib/utils/log.js:29:15)
    at Object.error (/workspace/mgit2/lib/utils/log.js:24:18)
    at execCommand.execute.then.then.then.then.then.then.then.catch (/workspace/mgit2/lib/commands/update.js:93:10)
    at process._tickCallback (internal/process/next_tick.js:103:7)
```

There are two issues here:

1. That I needed to modify this code to get any meaningful error because, by default, you only see a short info about unhandled promise.
2. That our `git status -s` doesn't catch this particular information. However, if the error which I can see inside `catch()` in `update.js` after adding `console.log()` there, was properly logged to the console it would be enough. After all, the situation I have (with unfinished rebase) is not very common.

### LINKED ISSUE #52: Introduce mgit checkout
[Bisecting using `mgit update`](https://github.com/ckeditor/ckeditor5/wiki/Development-environment#bisecting-through-a-multi-repo) is slooow. `mgit checkout` would help.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced new commands and fixes bugs related to invalid displayed errors. Closes #2. Closes #45. Closes #49. Closes #52.

- New commands:
  * `checkout` – changes branches in repositories according to the configuration file (see #52),
  * `diff` – prints changes from all repositories (see #2),
  * `status` – prints a table which contains useful information about the status of repositories (see #2).
- Bug fixes:
  * known errors should not be logged as crashes (see #45),
  * fixed `Unhandled promise` error (see #49).
- Added support for aliases (`st` => `status`, `co` => `checkout`).

---

### Additional information

On the occasion, I simplified the promises flow in the commands.

### CODE REVIEW

**Comment by pomek** [comment]
image

image

So we need to restart the build after some time.

File: README.md [hunk 1]
```diff
@@ -98,6 +98,12 @@ CLI options:
 --scope                     Restricts the command to packages which names match the given glob pattern.
 
                             Default: null
+                            
+--packages-prefix           The common name of the packages. The prefix will be removed from packages' names in order to
```
Comment (Reinmar) [inline README.md]: Why is this a CLI config option? Do you expect anyone to use it? I don't. If anywhere – this is for `mgit.json` only.

**Comment by Reinmar** [comment]
Since you're adding the `checkout` command, I think that we should also clean up the bootstrap/update mess. It would be good to decide what commands we use and how and when.

File: README.md (hunk 1 above)
Comment (pomek) [inline README.md]: Removed in d95291a.

**Comment by pomek** [comment]
Could we clean up the mess after merging these changes? I would like to avoid bigger PR than we have now. After merging we will be able to decide how these commands should work and what should do.

**Comment by Reinmar** [comment]
Yes, we can. But if you wanted to add `mgit status` we should've bootstrap this discussion before. DOn't dive into topics which are not yet clarified because you may be wasting your time.

File: README.md [hunk 2]
```diff
@@ -255,6 +254,62 @@ Example:
 mgit save-hashes
 ```
 
+### status
+##### Available also `st`
```
Comment (Reinmar) [inline README.md]: That shouldn't be a heading. Don't use semantical markup for styling reasons.

File: README.md [hunk 3]
```diff
@@ -255,6 +254,62 @@ Example:
...
+```bash
+mgit status 
+# or 
+mgit st
+```
+
+In order to save space in your terminal, you can define the `packagesPrefix` option in your configuration file.
+The prefix will be removed from packages' names. Full names of packages aren't needed so we can cat the names.
+
+### diff
```
Comment (Reinmar) [inline README.md]: Please add a screenshot.

File: README.md (hunk 2 above)
Comment (Reinmar) [inline README.md]: We can have `### stats (alias: st)`.

Please also add a screenshot.

File: lib/commands/bootstrap.js [hunk 4]
```diff
@@ -21,57 +21,55 @@ module.exports = {
 	execute( data ) {
 		const log = require( '../utils/log' )();
 
-		return new Promise( ( resolve, reject ) => {
-			const destinationPath = path.join( data.options.packages, data.repository.directory );
+		const destinationPath = path.join( data.options.packages, data.repository.directory );
```
Comment (Reinmar) [inline lib/commands/bootstrap.js]: Did you change anything in this file or is it just a code style?

Comment (Reinmar) [inline lib/commands/bootstrap.js]: If so – why changing anything?

File: lib/commands/savehashes.js [hunk 5]
```diff
@@ -18,29 +18,22 @@ module.exports = {
 		const log = require( '../utils/log' )();
 		const execCommand = require( './exec' );
 
-		return new Promise( ( resolve, reject ) => {
-			execCommand.execute( getExecData( 'git rev-parse HEAD' ) )
-				.then( execResponse => {
-					const commitHash = execResponse.logs.info[ 0 ];
+		return execCommand.execute( getExecData( 'git rev-parse HEAD' ) )
```
Comment (Reinmar) [inline lib/commands/savehashes.js]: The same question as about `bootstrap`.

File: lib/commands/update.js [hunk 6]
```diff
@@ -22,82 +22,77 @@ module.exports = {
 		const bootstrapCommand = require( './bootstrap' );
 		const execCommand = require( './exec' );
 
-		return new Promise( ( resolve, reject ) => {
-			const destinationPath = path.join( data.options.packages, data.repository.directory );
+		const destinationPath = path.join( data.options.packages, data.repository.directory );
```
Comment (Reinmar) [inline lib/commands/update.js]: The same – what are these changes?

**Comment by Reinmar** [comment]
I found one issue:

image

**Comment by Reinmar** [comment]
And the `co` task could print less info if everything went well. This is hard to parse:

image

One line per repo would be better. Something like "HEAD is now at f1131bc... Merge pull request #33 from ckeditor/t/32" will be fine.

**Comment by Reinmar** [comment]
Another bug – `mgit update` leaves repos on detached state. State before and after calling `mgit update`:

image

This is a regression.

**Comment by pomek** [comment]
> I found one issue:

Ok, mgit should not show staged files as modified also. I'll correct it.

File: lib/commands/bootstrap.js (hunk 4 above)
Comment (pomek) [inline lib/commands/bootstrap.js]: Changes in `bootstrap`, `savehashes` and `update` will resolve #45 and #49.

**Comment by pomek** [comment]
> And the co task could print less info if everything went well. This is hard to parse:

It is also hard to predict what we want to display:

image

Maybe should we display the last line of the log instead of all?

**Comment by Reinmar** [comment]
> Maybe should we display the last line of the log instead of all?

👍

**Comment by pomek** [comment]
image

**Comment by Reinmar** [comment]
I've been using these new commands for some time and they are great. No more issues found.
